openshift/endpoints.go

```diff
@@ -0,0 +1,204 @@
+
+		environment.ValKindDeploymentConfig: endpoints(
+			endpoint(`/apis/apps.openshift.io/v1/namespaces/{{ index . "metadata" "namespace"}}/deploymentconfigs`, POST(WhenConflictThenDeleteAndRedo)),
+			endpoint(`/apis/apps.openshift.io/v1/namespaces/{{ index . "metadata" "namespace"}}/deploymentconfigs/{{ index . "metadata" "name"}}`, PATCH(), GET(), DELETE())),
+
+		environment.ValKindPersistenceVolumeClaim: endpoints(
+			endpoint(`/api/v1/namespaces/{{ index . "metadata" "namespace"}}/persistentvolumeclaims`, POST(IgnoreConflicts)),
+			endpoint(`/api/v1/namespaces/{{ index . "metadata" "namespace"}}/persistentvolumeclaims/{{ index . "metadata" "name"}}`, PATCH(), GET(), DELETE())),
+
+		environment.ValKindService: endpoints(
+			endpoint(`/api/v1/namespaces/{{ index . "metadata" "namespace"}}/services`, POST(WhenConflictThenDeleteAndRedo)),
+			endpoint(`/api/v1/namespaces/{{ index . "metadata" "namespace"}}/services/{{ index . "metadata" "name"}}`, PATCH(), GET(), DELETE())),
+
+		environment.ValKindSecret: endpoints(
+			endpoint(`/api/v1/namespaces/{{ index . "metadata" "namespace"}}/secrets`, POST(WhenConflictThenDeleteAndRedo)),
+			endpoint(`/api/v1/namespaces/{{ index . "metadata" "namespace"}}/secrets/{{ index . "metadata" "name"}}`, PATCH(), GET(), DELETE())),
+
+		environment.ValKindServiceAccount: endpoints(
+			endpoint(`/api/v1/namespaces/{{ index . "metadata" "namespace"}}/serviceaccounts`, POST(IgnoreConflicts)),
+			endpoint(`/api/v1/namespaces/{{ index . "metadata" "namespace"}}/serviceaccounts/{{ index . "metadata" "name"}}`, PATCH(), GET(), DELETE())),
+
+		environment.ValKindConfigMap: endpoints(
+			endpoint(`/api/v1/namespaces/{{ index . "metadata" "namespace"}}/configmaps`, POST(WhenConflictThenDeleteAndRedo)),
+			endpoint(`/api/v1/namespaces/{{ index . "metadata" "namespace"}}/configmaps/{{ index . "metadata" "name"}}`, PATCH(), GET(), DELETE())),
+
+		environment.ValKindResourceQuota: endpoints(
+			endpoint(`/api/v1/namespaces/{{ index . "metadata" "namespace"}}/resourcequotas`, POST(WhenConflictThenDeleteAndRedo, GetObject)),
+			endpoint(`/api/v1/namespaces/{{ index . "metadata" "namespace"}}/resourcequotas/{{ index . "metadata" "name"}}`, PATCH(), GET(), DELETE())),
+
+		environment.ValKindLimitRange: endpoints(
+			endpoint(`/api/v1/namespaces/{{ index . "metadata" "namespace"}}/limitranges`, POST(WhenConflictThenDeleteAndRedo)),
+			endpoint(`/api/v1/namespaces/{{ index . "metadata" "namespace"}}/limitranges/{{ index . "metadata" "name"}}`, PATCH(), GET(), DELETE())),
+
+		environment.ValKindJob: endpoints(
+			endpoint(`/apis/batch/v1/namespaces/{{ index . "metadata" "namespace"}}/jobs`, POST(WhenConflictThenDeleteAndRedo)),
+			endpoint(`/apis/batch/v1/namespaces/{{ index . "metadata" "namespace"}}/jobs/{{ index . "metadata" "name"}}`, PATCH(), GET(), DELETE())),
+	}
```

Comment:
@MatousJobanek one of the basic/major design feedback I have on `endpoint` object.

I think the _old arrangement_ was simple [code_link](https://github.com/fabric8-services/fabric8-tenant/blob/master/openshift/apply.go#L28) **where endpoints are grouped for given action** (POST/PATCH)
rather, this _new arrangement_ **where endpoints are tagged with action it supports**.

In old_code, I really don't like `recursive` call to apply() [code_link](https://github.com/fabric8-services/fabric8-tenant/blob/master/openshift/apply.go#L233) which was making it hard/complex to understand.

In new_code, I like the way code is reused :-)

old one goes like this (in the code has definition of sequence_flow) .. if you want to `Setup()` tenant, `POST` is the action and for POST these are list_of_endpoints.  Call them one by one and response of each endpoint has to be further examined to take appropriate steps .. this helps in debugging the issue. for ex, we got issue which say when I provision_tenant, I get so and so error.  To debug just open code and see what are the endpoints called for `POST` and follow the execution of each endpoint.

So overall, see if you can fix endpoint_arrangement while still keep rest of design which provides code_reusability.  If it's a lot_work than you can take one more feedback from team and see if other feel the this problem or not.

Reply:
@nurali-techie thank you very much for your feedback and sorry for may delayed response.

> I think the old arrangement was simple code_link where endpoints are grouped for given action (POST/PATCH) rather, this new arrangement where endpoints are tagged with action it supports.

I see your point - I think that it is just a question of the point of view. The old code was grouped by the action/method, but was sometimes hard to find all endpoints for one object kind. And also to check that there are all endpoints set for all actions for a particular object kind. Now all information are grouped together and visible in one place.

> So overall, see if you can fix endpoint_arrangement while still keep rest of design which provides code_reusability. If it's a lot_work then you can take one more feedback from team and see if other feel the this problem or not.

I could do that, but the only thing that would be changed is the endpoints arrangement. The internal logic would stay the same (because I rely on it in other places) so I'm not sure if that would anyhow help with debugging. I'm not sure how much time that would need, maybe a few hours, maybe one day. Let's wait for the other guys what they think about it.